marshal/marshal.go

```diff
@@ -28,91 +30,371 @@ const (
 	HeartbeatInterval = 60 // Measured in seconds.
 )
 
+// Marshaler is the coordinator type. It is designed to be used once globally and
+// is thread safe. Creating one of these will create connections to your Kafka
+// cluster and begin actively monitoring the coordination topic.
+type Marshaler struct {
+	// These members are not protected by the lock and can be read at any
+	// time as they're write-once or only ever atomically updated. They must
+	// never be overwritten once a Marshaler is created.
+	quit       *int32
+	cluster    *cluster
+	instanceID string
+	clientID   string
+	groupID    string
+	offsets    kafka.OffsetCoordinator
+
+	// Lock protects the following members; you must have this lock in order to
+	// read from or write to these.
+	lock      sync.RWMutex
+	consumers []*Consumer
+}
+
 // NewMarshaler connects to a cluster (given broker addresses) and prepares to handle marshalling
 // requests. Given the way this system works, the marshaler has to process all messages in the
 // topic before it's safely able to begin operating. This might take a while.
 func NewMarshaler(clientID, groupID string, brokers []string) (*Marshaler, error) {
-	// TODO: It might be nice to make the marshaler agnostic of clients and able to support
-	// requests from N clients/groups. For now, though, we require instantiating a new
-	// marshaler for every client/group.
-	brokerConf := kafka.NewBrokerConf("PortalMarshal")
-	broker, err := kafka.Dial(brokers, brokerConf)
+	cluster, err := getOrCreateCluster(brokers)
```

Comment:
Ah okay, so this is how `example.go` didn't change at all. You still end up creating one `Marshaler` per (cluster, group) but under the hood Marshal just does the right thing. And the price is the global variables in `cluster.go` it looks like.